### DISCUSSION TITLE
Docker Build Image

### DISCUSSION THREAD

**Discussion by crempp** [open-watcom/open-watcom-v2#471] [Announcement]
I absolutely love this project and am so appreciative of all the work that has gone into it. I hope my minor contribution can help others.

The largest problem I've had with Open Watcom is setting up the build environment, especially on a Mac. Once it's setup it's great. For this reason I thought it may be helpful to have a Docker build image. Not only will this make it incredibly easy to get started using Watcom but also it makes it possible/simple to use CI build tools like CircleCI or Jenkins.

Here's my initial stab at a Docker build image. It functions for my particular use case but I know I'm missing things and there's more work to do. I hope others will try it out, let me know what doesn't work and perhaps contribute.

https://hub.docker.com/r/lapinlabs/watcom/

**Comment by jmalak** [comment]
Hi, Thanks for your info.
The main problem with CI is that jobs are limited by run time, probably the same problem with sinilar sites.
On CI is possible to run build also for MAC, but it is always stopped due to maximum run time is overflow that I stopped it already.
To fix this issue build is divided to several jobs to not overflow maximum run time per job and use CI cache to store intermediate things to be available for each serialized job.
Build requires several jobs serialized:
initialization phase (extra job) - prepare GITHub repository
bootstrap phase (extra job)
build phase (extra job per executables + extra job for documentation)
test phase (extra job for each test (C compiler, C++ compiler, F77 compiler, Assembler, C run-time )
finalization phase (extra job) - update GITHub repository

If you have some trouble then you can try to ask. If I know than I will help you.

**Comment by crempp** [comment]
@jmalak I may not have explained this clearly.

This is Docker image is not to build Open Watcom itself. It's meant to be used for end users like myself to build my own project **using** Open Watcom.

For example. I'm writing a 16 bit OS and using Open Watcom v2 to compile. Rather than have to setup the Open Watcom environment on my local machine I can use this image. Here's a simple example. I use Scons but you can swap out for `make` pretty easily.

In my OS project I add a file `Dockerfile` containing
```
FROM lapinlabs/watcom

RUN apk add --no-cache --update scons

WORKDIR /app

CMD ["scons"]
```

Then I can build my project by 

1. Build the image (only need to do this once)
```
$ docker build -t my-builder .
```
2. Run 
```
$ docker run --rm -v `pwd`:/app my-builder
```

By doing this I can build my project by running that second command. I never had to setup Open Watcom on my local machine.

**Comment by jmalak** [comment]
I understand now.
May be interesting this Wiki page about [OW configuration for target platform](https://wiki.archlinux.org/index.php/Open_Watcom)

**Comment by mcejp** [comment]
For some reason, your Dockerfile didn't work for me in 2021. I [had to use](https://github.com/mcejp/watcom-docker/commit/fa749d3bc6af7885513876fbb26c903ccf01bcad#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557R25) this hack:
```
    # Comment out inclusion of wgml/builder.ctl to avoid Dosbox dependency -- this SHOULD NOT be necessary
    && sed '/\[ INCLUDE <OWSRCDIR>\/wgml\/builder.ctl \]/{s/^/#/}' -i bld/builder.ctl \
```

**Reply by jmalak** [reply]
wgml build is disabled now

**Reply by crempp** [reply]
Thanks for pointing this out. I'll take a look when I get a moment.

**Reply by mcejp** [reply]
Ok, 6eda89da37e34e0c3b00b143189d9b4c81650869 fixes this issue -- thanks, Jiří!
Chad, I have more feedback for your Docker image if you're interested, but I'm not sure if it belongs in this thread or somewhere else.
